drivers/spi/spi_pl022.c

```diff
@@ -0,0 +1,499 @@
+	cr0 |= SSPCR0_SCR(SCR1 - 1);		/* Serial clock rate */
+	if (spi_config->config & SPI_MODE_CPOL) {
+		cr0 |= SSPCR0_SPO;		/* Clock polarity */
+	}
+	if (spi_config->config & SPI_MODE_CPHA) {
+		cr0 |= SSPCR0_SPH;		/* Clock phase */
+	}
+	cr0 |= SSPCR0_DSS(context->data_bits - 1); /* Data bits */
+
+	cr1 = 0;
+	if (spi_config->config & SPI_MODE_LOOP) {
+		cr1 |= SSPCR1_LBM;		/* Loopback */
+	}
+	cr1 |= SSPCR1_SSE;			/* Enable port */
+
+	ssp->cr0 = cr0;
+	ssp->cr1 = cr1;
+
+	return 0;
+}
+
+/*
+ * Define a programmed I/O function optimised for a given data size (8/16-bit)
+ * and whether we want to transmit data, receive data, or do both.
+ */
+#define DEFINE_PIO_FUNCTION(_name, _size, _tx, _rx)			\
+									\
+static void _name##_size(struct spi_pl022_context *context,		\
+			struct spi_pl022 *ssp)				\
+{									\
+	uint##_size##_t * ptr; /* Wrong spacing to pass checkpatch */	\
+	unsigned int remain;						\
+	unsigned int limit;						\
+									\
+	ptr = (uint##_size##_t *)context->rx_buf;			\
+	remain = context->rx_remain;					\
+	while ((ssp->sr & SSPSR_RNE) && remain) {			\
+		if (_rx) {						\
+			*ptr++ = ssp->dr;				\
+		} else {						\
+			ssp->dr; /* Discard unwanted received data */	\
+		}							\
+		--remain;						\
+	}								\
+	context->rx_buf = (u8_t *)ptr;				\
+	context->rx_remain = remain;					\
+									\
+	/* limit transmission so it can't overflow receive FIFO */	\
+	limit = remain > FIFO_DEPTH ? remain - FIFO_DEPTH : 0;		\
+									\
+	ptr = (uint##_size##_t *)context->tx_buf;			\
+	remain = context->tx_remain;					\
+	while ((ssp->sr & SSPSR_TNF) && remain > limit) {		\
+		ssp->dr = _tx ? *ptr++ : 0;				\
+		--remain;						\
+	}								\
+	context->tx_buf = (u8_t *)ptr;				\
+	context->tx_remain = remain;					\
+}
+
+DEFINE_PIO_FUNCTION(_transceive, 8, true, true)
+DEFINE_PIO_FUNCTION(_transceive, 16, true, true)
+DEFINE_PIO_FUNCTION(_transmit, 8, true, false)
+DEFINE_PIO_FUNCTION(_transmit, 16, true, false)
+DEFINE_PIO_FUNCTION(_receive, 8, false, true)
+DEFINE_PIO_FUNCTION(_receive, 16, false, true)
+
+/*
+ * Interrupt service routine called for all interrupt types the PL022 can raise.
+ */
+void spi_pl022_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct spi_pl022_context *context = dev->driver_data;
+	struct spi_pl022 *ssp = context->ssp;
+
+	/* Transfer as much data as FIFO permits */
+	context->transfer_fn(context, context->ssp);
+
+	if (!context->rx_remain) {
+		/* Received everything, so we're done... */
+		ssp->imsc = 0; /* Disable all our interrupts */
+		k_sem_give(&context->transfer_sem);
+	} else if (!context->tx_remain) {
+		/*
+		 * Nothing more to send, so disable tx interrupt
+		 * and enable receive timeout interrupt so we get notified
+		 * after last item of data has been received.
+		 */
+		ssp->imsc = SSP_INT_RT;
+	}
+}
+
+/*
+ * Try and sleep task/fibre for a period of time (in microseconds)
+ */
+static void idle_for(unsigned int usec)
+{
+	if (k_is_in_isr())
+		return;
```

Comment:
OK, I'll delete polled mode. But for a lot of use cases with high SPI clock rates the system will permanently be in the ISR as the CPU can't keep the FIFO filled even with a "while(count--) *fifo++ = *data++;". So, having the hundreds of cycles overhead from an ISR routine will just reduce the data rate by a factor of 10 for no benefit. And having the ISR being permanently triggered will block other tasks, whereas if the polling mode is happening in the client thread context, other tasks of higher priority get to run.

Also, last time I checked, ARM's simulations (Fixed Virtual Platforms) didn't implement ISRs (or were buggy) so by deleting polled I would stop Zephy being able to display anything on the screen (which is attached by SPI). So I guess I'll keep the polled mode as an out-of-tree patch.